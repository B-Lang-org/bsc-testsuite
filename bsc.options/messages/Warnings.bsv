// Generate four warnings:
//
// (1) Arbitrary earliness warning (G0036)
// (2) Action shadowing warning (G0117)
// (3) Arbitrary urgency warning (G0010)
// (4) System task in method warning (G0020)
//

interface Ifc;
   method Action m1();
   method Action m2();
endinterface

(* synthesize *)
module sysWarnings (Ifc);
   Reg#(Bool) rg1 <- mkRegU;
   Reg#(Bool) rg2 <- mkRegU;
   Reg#(Bool) rg3 <- mkRegU;

   // These rules have arbitrary earliness, and one shadows the other
   rule r1;
      rg1 <= True;
   endrule
   rule r2;
      rg1 <= False;
   endrule

   // This rule and method have arbitrary urgency
   rule r3;
      rg2 <= rg3;
   endrule
   method Action m1();
      rg3 <= rg2;
   endmethod

   // This method contains a system task
   method Action m2();
      $display("m2");
   endmethod
endmodule

