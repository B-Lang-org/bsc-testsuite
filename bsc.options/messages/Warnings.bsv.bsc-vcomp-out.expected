checking package dependencies
compiling Warnings.bsv
code generation for sysWarnings starts
Warning: "Warnings.bsv", line 15, column 8: (G0020)
  System functions (e.g., $display) called by interface methods execute in an
  unpredictable order during Verilog simulations.
      Top-level interface method `m2' calls a system function (e.g., $display)
      at "Warnings.bsv", line 38, column 7
Warning: "Warnings.bsv", line 15, column 8: (G0010)
  Rule "m1" was treated as more urgent than "r3". Conflicts:
    "m1" cannot fire before "r3": calls to rg3.write vs. rg3.read
    "r3" cannot fire before "m1": calls to rg2.write vs. rg2.read
Warning: "Warnings.bsv", line 15, column 8: (G0036)
  Rule "r1" will appear to fire before "r2" when both fire in the same clock
  cycle, affecting:
    calls to rg1.write vs. rg1.write
Warning: "Warnings.bsv", line 15, column 8: (G0117)
  Rule `r2' shadows the effects of `r1' when they execute in the same clock
  cycle. Affected method calls:
    rg1.write
  To silence this warning, use the `-no-warn-action-shadowing' flag.
Verilog file created: sysWarnings.v
All packages are up to date.
